modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/OptimizingPhaseHandler.java

```diff
@@ -0,0 +1,151 @@
+        validateDynamicParameters(result.dynamicParamsCount(), query.params, true);
+        ensureStatementMatchesTx(result.queryType(), query.txContext);
+
+        HybridTimestamp operationTime = query.executor.deriveOperationTime(query.txContext);
+
+        String schemaName = query.properties.get(QueryProperty.DEFAULT_SCHEMA);
+        ZoneId timeZoneId = query.properties.get(QueryProperty.TIME_ZONE_ID);
+
+        SqlOperationContext operationContext = SqlOperationContext.builder()
+                .queryId(query.id)
+                .cancel(query.cancel)
+                .prefetchCallback(new PrefetchCallback())
+                .parameters(query.params)
+                .timeZoneId(timeZoneId)
+                .defaultSchemaName(schemaName)
+                .operationTime(operationTime)
+                .txContext(query.txContext)
+                .txUsedListener(tx -> query.usedTransaction = tx)
+                .build();
+
+        query.operationContext = operationContext;
+
+        CompletableFuture<Void> awaitFuture = query.executor.waitForMetadata(operationTime)
+                .thenCompose(none -> query.executor.prepare(result, operationContext)
+                        .thenAccept(plan -> {
+                            if (query.txContext.explicitTx() == null) {
+                                // in case of implicit tx we have to update observable time to prevent tx manager to start
+                                // implicit transaction too much in the past where version of catalog we used to prepare the
+                                // plan was not yet available
+                                query.txContext.updateObservableTime(query.executor.deriveMinimalRequiredTime(plan));
+                            }
+
+                            query.plan = plan;
+
+                            query.moveTo(ExecutionPhase.CURSOR_INITIALIZATION);
+                        }));
+
+        return Result.proceedAfter(awaitFuture);
+    }
+
+    /** Performs additional validation of a parsed statement. **/
+    private static void validateParsedStatement(
+            SqlProperties properties,
+            ParsedResult parsedResult
+    ) {
+        Set<SqlQueryType> allowedTypes = properties.get(QueryProperty.ALLOWED_QUERY_TYPES);
+        SqlQueryType queryType = parsedResult.queryType();
+
+        if (parsedResult.queryType() == SqlQueryType.TX_CONTROL) {
+            String message = "Transaction control statement can not be executed as an independent statement";
+
+            throw new SqlException(STMT_VALIDATION_ERR, message);
+        }
+
+        if (!allowedTypes.contains(queryType)) {
+            String message = format("Invalid SQL statement type. Expected {} but got {}", allowedTypes, queryType);
+
+            throw new SqlException(STMT_VALIDATION_ERR, message);
+        }
+    }
+
+    private static void validateDynamicParameters(int expectedParamsCount, Object[] params, boolean exactMatch) throws SqlException {
```

Comment:
Looks like this function is written 3 times - here, in the `SqlQueryProcessor` and in the `MultistatementHandler`).
Perhaps it would be better to have a single definition.

Reply:
good catch! Moved all such validation to `ValidationHelper` in order to reuse it